### PR TITLE
GPT-OSS: fix 120B router-weights test via realistic dummy weight scale

### PR DESCRIPTION
**Pull request by mtairum**
## Summary
Fixes the GPT-OSS **120B router-weights** unit-test divergence. Split out of the transformers-5.10.2 unpin (#47671, now merged); rebased onto `main`. **No kernel change** — the fix is a test-only correction.

## Root cause — a dummy-weight test artifact, not a real precision bug
`setup_reference_layer` in `tests/unit/test_modules.py` initialized the layer's router/experts/sinks at **unit-normal (`normal_(0, 1)`, std=1)**. The router logit is `hidden @ W_router`, so with std=1 weights the logit std is `≈ sqrt(hidden_size) ≈ 54` → router logits reach **~±150**.

At that magnitude the **bf16-only `ttnn.topk`** ties adjacent top logits (e.g. `109.0 == 109.0`), so the gate softmax flattens to ~**0.5/0.5** where the fp32 reference gives ~**0.95/0.05** — collapsing the gate PCC to **0.6157**. Same experts are selected (indices PCC 1.0); only the gate magnitudes diverged.

**The real model never sees these magnitudes.** GPT-OSS `config.initializer_range = 0.02`, which gives router logits ~**±3** — comfortably within bf16 resolution. The "bug" was purely an artifact of the unrealistic random-weight scale used in the test.

> The earlier revision of this PR fixed the symptom by forcing the router gate to **fp32** (`topk.py` row-max-shift + HiFi4). That has been **reverted**: fp32 in the router is undesirable, and unnecessary once the test uses realistic weights. (Thanks to review feedback for the steer.)

## Changes
- **`tests/unit/test_modules.py` — initialize dummy weights at the real scale:** `setup_reference_layer` now uses `normal_(0, config.initializer_range)` (0.02) instead of `normal_(0, 1)`. Real-scale logits (~±3) stay within bf16 resolution, so the gate matches the fp32 reference with no kernel change.
- **`tests/unit/test_modules.py` — corrected router-weights comparison (kept):** gather TT's dense non-throughput weights at the selected expert ids (instead of the first `top_k` unselected columns) and normalize the reference to the same layout. This makes the `test_decoder[decode_low_latency]` sub-check actually validate the gate (it previously false-passed by reading unselected columns). Runs in the tier-1/2 unit pipeline.

`topk.py` and `tests/fused_op_unit_tests/test_gpt_oss_router.py` are **unchanged from main** (the fp32 fix and its showcase test are removed).

## Validation
**[Run 28178109629](https://github.com/tenstorrent/tt-metal/actions/runs/28178109629)** (tier-1 120B + tier-2 20B unit, original topk + realistic weights):
- **120B [wh_llmbox] → ✅** · **120B [bh_quietbox_2] → ✅** · **120B [bh_galaxy] → ✅** — `test_decoder[decode_low_latency]` passes (was 0.6157).
- **20B [wh_llmbox] → ✅** · **20B [bh_quietbox_2] → ✅** — no regression.
- **20B [bh_p150] → ✅** (clean [rerun 28190842155](https://github.com/tenstorrent/tt-metal/actions/runs/28190842155); the original attempt infra-flaked on a GitHub codeload action-download timeout).
- wh_galaxy disregarded (chronically flaky/infra, per maintainer guidance).

The corrected comparison now passes on all healthy SKUs **with no kernel change**, confirming the divergence was a test-weight-scale artifact.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mtairum/gptoss-router-fp32-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mtairum/gptoss-router-fp32-gate)

### CODE REVIEW

File: models/demos/gpt_oss/tests/fused_op_unit_tests/test_gpt_oss_router.py [hunk 1]
```diff
@@ -867,6 +869,91 @@ def test_gpt_oss_router_single_device(
...
+    # logits whose gaps (1.0) are below bf16's step at this magnitude (~512 -> step ~4), so the pre-fix bf16
+    # topk ties/flattens them while the fp32 reference keeps a peaked softmax.
+    bias = torch.full((num_experts,), -64.0, dtype=torch.float32)
+    base = 512.0
+    for i in range(top_k):
+        bias[i] = base - float(i)
+    state_dict = {
+        "weight": torch.zeros(num_experts, hidden_size, dtype=torch.float32),
+        "bias": bias.clone(),
+    }
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/demos/gpt_oss/tests/fused_op_unit_tests/test_gpt_oss_router.py]: Good catch — fixed in 6cb3d18c9fd. You're right: `bias` is stored as bf16 in `TopKRouter`, so the crafted gaps were rounded away before the fp32 path ever ran, and the test would not have passed even with the fix.

The test now drives the large, near-tied logits **through the matmul** (all-ones hidden × a crafted weight whose top-k rows sum to [512, 509, 506, 503], zero bias), and the fp32 reference is computed from the **same bf16-stored weight** (`weight.to(bfloat16).float().sum(dim=1)`) to match the TT HiFi4/fp32-accumulate matmul. So the gaps now live in the fp32 matmul accumulation + row-max shift: it fails pre-fix (large logits cast to bf16 before topk → tied gate) and passes post-fix.

**Comment by mtairum** [comment]
## ✅ All-SKU validation — router-gate fix confirmed (wh_galaxy flaky, unrelated)

Ran the full SKU matrix for both models on HEAD. **The fix passes on every healthy SKU:**

| Model | SKU | Result |
|-------|-----|--------|
| 120B | wh_llmbox | ✅ |
| 120B | bh_quietbox_2 | ✅ (×2) |
| 120B | bh_galaxy | ✅ |
| 120B | **wh_galaxy** | ❌ flaky/infra — see below |
| 20B | wh_llmbox | ✅ |
| 20B | bh_p150 | ✅ |
| 20B | bh_quietbox_2 | ✅ (green on retry) |

`decode_low_latency` — the router-weights test this PR fixes — passes on **all** of the above (including wh_galaxy itself in the first run).

### wh_galaxy is flaky/infra, not this fix
wh_galaxy failed twice with **two different, fix-unrelated** modes:
- Run [28158751749](https://github.com/tenstorrent/tt-metal/actions/runs/28158751749): `test_decoder[...decode_high_throughput-4x8]` → `IndexError` inside `run_fused_throughput_experts_component` (the HF reference experts forward). **This PR does not touch that function, and it passes on `main`** (same code, same transformers). It builds routing/hidden with **unseeded `torch.randperm`/`torch.randn`**, so it's nondeterministic — worth **seeding separately** (pre-existing, out of scope here).
- Run [28165275397](https://github.com/tenstorrent/tt-metal/actions/runs/28165275397): job **hung** (no pytest summary, hang-triage fired) → pure infra/device instability.

Different failure each time on the same SKU = chronic wh_galaxy device instability (consistent with this SKU's behavior throughout this work). Not gating on it; the router-gate fix is validated on every other SKU.

**Comment by mtairum** [comment]
## Approach change — drop the fp32 gate, fix the test's weight scale instead

After review feedback: the bf16 router-gate "precision bug" this PR was fixing is a **test artifact, not a real-model issue**, and forcing the router gate to fp32 is undesirable. So this PR is reworked:

**Root cause (corrected):** `setup_reference_layer` in `test_modules.py` initialized router/experts/sinks at **unit-normal (std=1)**. The GPT-OSS router logit is `hidden @ W_router`, so with std=1 weights the logit std is `≈ sqrt(hidden_size) ≈ 54` → logits reach **~±150**. At that magnitude the bf16-only `ttnn.topk` ties adjacent top experts and the gate softmax collapses to ~0.5/0.5 vs the fp32 reference's ~0.95/0.05. The real model never sees these magnitudes: GPT-OSS `config.initializer_range = 0.02`, which gives logits **~±3** — comfortably within bf16 resolution.

**Fix:** initialize the dummy weights at the real scale (`config.initializer_range`) instead of std=1. The corrected router-weights comparison then passes with **no kernel change**.

**Changes vs the previous revision:**
- ❌ Reverted `topk.py` fp32 row-max-shift gate + HiFi4/fp32 compute config (no longer needed).
- ❌ Dropped the `test_gpt_oss_router_gate_fp32_precision` showcase test (it demonstrated the now-removed fp32 path).
- ✅ Kept the corrected router-weights comparison (reads the actual selected-expert gate weights).
- ✅ `setup_reference_layer`: `normal_(0, 1)` → `normal_(0, config.initializer_range)`.

Net diff is now a single test file. Rebased on latest `main`. Validation run (20B + 120B, tiers 1+2, all SKUs except the chronically-flaky `wh_galaxy`): https://github.com/tenstorrent/tt-metal/actions/runs/28178109629

**Comment by mtairum** [comment]
✅ **Validation complete — all healthy SKUs green** (original topk + realistic weights, no kernel change):
- **120B:** wh_llmbox ✅ · bh_quietbox_2 ✅ · bh_galaxy ✅
- **20B:** wh_llmbox ✅ · bh_quietbox_2 ✅ · bh_p150 ✅ (clean [rerun](https://github.com/tenstorrent/tt-metal/actions/runs/28190842155); original attempt hit a GitHub codeload action-download timeout)
- wh_galaxy disregarded (chronically flaky/infra).

The corrected router-weights comparison passes everywhere once the dummy weights are initialized at the real `initializer_range` scale, confirming the bf16 gate-tie was a test-weight-scale artifact rather than a model precision bug. Ready for review.

**Review by handrewsTT — APPROVED** [review]
